csr: correctly handle backdating of short lived certs

enj opened the pull request:
This change updates the backdating logic to only be applied to the
NotBefore date and not the NotAfter date when the certificate is
short lived. Thus when such a certificate is issued, it will not be
immediately expired.  Long lived certificates continue to have the
same lifetime as before.

Consolidated all certificate lifetime logic into the
PermissiveSigningPolicy.policy method.

Signed-off-by: Monis Khan <mok@vmware.com>

/kind bug
/sig auth
/priority important-longterm
/triage accepted
/milestone v1.22
@kubernetes/sig-auth-pr-reviews 

```release-note
NONE
```

enj commented:
/assign @mikedanese @liggitt

enj commented on pkg/controller/certificates/authority/authority.go:
```diff
@@ -78,16 +68,19 @@ func (ca *CertificateAuthority) Sign(crDER []byte, policy SigningPolicy) ([]byte
-	if !now.Before(ca.Certificate.NotAfter) {
+	if !time.Now().Before(ca.Certificate.NotAfter) {
 		return nil, fmt.Errorf("refusing to sign a certificate that expired in the past")
 	}
```
If we want to keep this check, I think I should pass in the CA into the `SigningPolicy.apply` call (to keep all the `now` logic in one place).

enj replied:
Or something like:

```go
type SigningPolicy interface {
	apply(template *x509.Certificate, usages []capi.KeyUsage, caNotAfter time.Time) error
}
```

mikedanese commented:
I can't decide if "backdate" here mean "pretend I did this earlier" or something else. Some SSL clients put a cap on the validity duration, e.g. https://support.apple.com/en-us/HT211025. Would it be surprising to request a certificate from a signer configured for 24 hours and get one back with validity for 24 hours + 5 minutes?

> Thus when a short lived certificate is issued, it will not be immediately expired.

Alternatively, we could not backdate NotAfter for certificates with TTLs of e.g. less than 30 minutes or an 5 minutes or an hour.

enj commented:
> I can't decide if "backdate" here mean "pretend I did this earlier" or something else

I always took backdating of a cert to mean "set the starting validity time to a little bit earlier than `now` to account for clock differences among consumers."

> Some SSL clients put a cap on the validity duration

I do not think we can account for all such restrictions and have to rely on something sane being passed to `--cluster-signing-duration`.  The one year default we have today at least falls within Apple's restrictions.

> Would it be surprising to request a certificate from a signer configured for 24 hours and get one back with validity for 24 hours + 5 minutes?

To be fair, consumers of the CSR API have no clue what they are going to get back today, which I think is a far worse issue.  I doubt anyone using the default one year duration cares if it is exactly one year or a few minutes off.  The shorter the cert lifetime, the more likely that someone is to care about the specifics.  If we assume that most consumers have similar clocks (which may be a big assumption), I think its safe to only backdate `notBefore` since that refers to a time in the past.  Those extra 5 minutes have likely already passed.

> we could not backdate NotAfter for certificates with TTLs of e.g. less than 30 minutes or 5 minutes or an hour.

The problem with that approach is that such clients are likely to be doing constant cert renewals since they are being issued short lived creds.  Thus they are more likely to encounter clock differences since their `notBefore` time is constantly moving forward.

desponda commented:
👋 Hello! I am from the Bug Triage team! Friendly reminder that code freeze is starting March 9th, 2021 (about 1 week from now). We want to ensure that each PR has a chance to be merged before code freeze.
As the PR is tagged for 1.21, is it still planned for this release?

enj commented:
> As the PR is tagged for 1.21, is it still planned for this release?

Yes.  Just waiting on some more review from Mike at this moment.

erismaster commented:
@enj We are past code freeze for 1.21, I am going to change this to the 1.22 milestone.

/milestone v1.22

enj committed:
csr: correctly handle backdating of short lived certs

This change updates the backdating logic to only be applied to the
NotBefore date and not the NotAfter date when the certificate is
short lived. Thus when such a certificate is issued, it will not be
immediately expired.  Long lived certificates continue to have the
same lifetime as before.

Consolidated all certificate lifetime logic into the
PermissiveSigningPolicy.policy method.

Signed-off-by: Monis Khan <mok@vmware.com>

liggitt commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -31,40 +31,76 @@ import (
-	// Usages are the allowed usages of a certificate.
-	Usages []capi.KeyUsage
+
+	// Backdate and Now are used to calculate the NotBefore value of the certificate.
```
it's a little confusing to only reference NotBefore on this field's doc... the PermissiveSigningPolicy doc was clearer.

enj replied:
https://github.com/kubernetes/kubernetes/pull/99412#discussion_r656605070

liggitt commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -31,40 +31,76 @@ import (
+	// Backdate and Now are used to calculate the NotBefore value of the certificate.
+	Backdate time.Duration
+
+	// Short is the duration used to determine if the lifetime of a certificate should be considered short.
```
might be helpful to repeat the "Short lived certificates ignore Backdate and only use TTL to compute NotAfter." bit in this doc.

liggitt replied:
would it be helpful to say something like this somewhere?
```
All certificates set NotBefore=Now()-Backdate.
Long-lived certificates set NotAfter=Now()+TTL-Backdate.
Short-lived certificates set NotAfter=Now()+TTL.
```

enj replied:
See what you think of the updated go docs.

liggitt commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -31,40 +31,76 @@ import (
 	tmpl.ExtKeyUsage = extUsages
-	tmpl.NotAfter = tmpl.NotBefore.Add(p.TTL)
+	tmpl.NotBefore = now.Add(-p.Backdate)
+	tmpl.NotAfter = now.Add(ttl - p.Backdate)
```
setting NotAfter and then resetting below is a little confusing.

if ttl < p.Short {
  tmpl.NotAfter = now.Add(ttl)
}else {
  now.Add(ttl - p.Backdate)
}

liggitt commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -31,40 +31,76 @@ import (
-	Usages []capi.KeyUsage
+
+	// Backdate and Now are used to calculate the NotBefore value of the certificate.
+	Backdate time.Duration
```
does anything make sure Backdate > TTL?

enj replied:
Based on the 10 minute minimum outlined in the CSR duration KEP, do you think [this comment](https://github.com/kubernetes/kubernetes/pull/99412/files#diff-13be1b8eeba9e7ea31ad3a75cee7bae2d0d396368c0baf184b0498e334c66a93R196) is sufficient?

liggitt commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -31,40 +31,76 @@ import (
 	tmpl.BasicConstraintsValid = true
 	tmpl.IsCA = false
 
+	if !tmpl.NotBefore.Before(signerNotAfter) {
+		return fmt.Errorf("the signer has expired: NotAfter=%v", signerNotAfter)
+	}
+
+	if !tmpl.NotAfter.Before(signerNotAfter) {
+		tmpl.NotAfter = signerNotAfter
+	}
+
+	if !now.Before(signerNotAfter) {
+		return fmt.Errorf("refusing to sign a certificate that expired in the past: NotAfter=%v", signerNotAfter)
+	}
```
if you have an expired signer, your life is not going to be great... but I'm trying to decide whether refusing to sign here makes life better or worse.

liggitt replied:
oh, just realized this is moving an existing expired check, nm...

liggitt commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -31,40 +31,76 @@ import (
-func (p PermissiveSigningPolicy) apply(tmpl *x509.Certificate) error {
-	usage, extUsages, err := keyUsagesFromStrings(p.Usages)
+func (p PermissiveSigningPolicy) apply(tmpl *x509.Certificate, usages []capi.KeyUsage, signerNotAfter time.Time) error {
+	now := time.Now()
+	if p.Now != nil {
+		now = p.Now()
+	}
```
setting and resetting is a little confusing

```go
var now time.Time
if p.Now != nil {
  now = p.Now()
} else {
  now = time.Now()
}
```

liggitt commented:
a few doc and code clarifications, lgtm otherwise

kikisdeliveryservice commented:
This has a pending exception request . Removing from milestone until exception is reviewed and KEP is finished.

/hold

liggitt commented:
> This has a pending exception request . Removing from milestone until exception is reviewed and KEP is finished.

this is actually a bugfix which is a prereq to the enhancement PRs but should be fixed whether or not the enhancement is done for 1.22

/hold cancel

kikisdeliveryservice commented:
> > This has a pending exception request . Removing from milestone until exception is reviewed and KEP is finished.
> 
> this is actually a bugfix which is a prereq to the enhancement PRs but should be fixed whether or not the enhancement is done for 1.22
> 
> /hold cancel

Thanks for clarifying @liggitt I will note this on the KEP issue.

enj commented:
/milestone v1.22

enj commented:
@liggitt addressed your comments here and also addressed the code structure comments from @mikedanese in #99494

liggitt commented on cmd/kubelet/app/server_bootstrap_test.go:
```diff
@@ -298,14 +298,16 @@ func (s *csrSimulator) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		}
 		der, err := ca.Sign(cr.Raw, authority.PermissiveSigningPolicy{
 			TTL: time.Hour,
+			Now: func() time.Time {
+				return time.Now().Add(-s.backdate)
+			},
```
should this be setting backdate instead:
```
Backdate: s.backdate,
```

enj replied:
I think this was leftover from the original code that changed the overall duration of all certs and not just short lived ones.  Fixed.

liggitt commented:
one question on a testcase that seemed to be doing baked-in backdating, lgtm otherwise

liggitt commented:
/lgtm
/approve
